### ISSUE TITLE
[JENKINS-39932] Support more arbitrary Pipeline definitions

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/pipeline-model-definition-plugin#896]
<p>I think the idea of a declarative (and not imperative) pipeline is a lot easier on the writers end (and I assume the implementers end as well), but I think the implementation of several serial stages each fanning out to a number of parallel stages is limiting to the overall description of the dependencies between different steps/nodes. In my case (simplifying, please excuse pseudo-code), I'd like to do something like this:</p>

<pre>
stage(<span class="code-quote">"checkout"</span>) { node () { git XXX } }
stage(<span class="code-quote">"build"</span>) {
    parallel (<span class="code-quote">"ubuntu"</span>) { node(<span class="code-quote">"ubuntu"</span>) { make XXX }},
 (<span class="code-quote">"centos"</span>) { node(<span class="code-quote">"centos"</span>) { make YYY }},
 <span class="code-comment">// tons more distributions here
</span>}
stage (<span class="code-quote">"test"</span>) {
    node (<span class="code-quote">"ubuntu"</span>) { test-ubuntu ZZZ }
}

stage (<span class="code-quote">"publish"</span>) { node() { publish.sh }
</pre>

<p>Since I know the "test" stage only depends on the "ubuntu" part of the build stage, I'd like the test stage to fire without waiting on the whole "build" stage completes. From what I see, the parallel block waits for everything inside to complete, but I know the dependencies and can speed it up (in my case, there are a lot more nodes who match the testing than who don't). If I were to write it in bash, I would do something like:</p>

<pre>
git checkout XXX
build ubuntu &
TEST_PID=$!
build centos &
wait $TEST_PID
test-ubuntu
wait # wait <span class="code-keyword">for</span> all testing and bulding to complete
publish
</pre>

<p>This is a rough way to explain the dependency graph to bash by saying "wait on 'build ubuntu' before starting the testing, but centos can take as long as it wants" and "before going to the publish step, make sure everything else is done". In my case, all the build steps for different distributions take forever because there's a finite number of VMs that can be spun up, but there's an abundance of "ubuntu" nodes that can be used for testing and compiling. It slows things down a ton to have to wait for the entire parallel node to finish before moving on, when there are free executors in latter steps that can start running.</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/perilousapricot.jpg" title="perilousapricot's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=perilousapricot">perilousapricot</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-39932" target="_blank">Support more arbitrary Pipeline definitions</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer">abayer</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>component(s)</b>: pipeline-model-definition-plugin
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2017-12-18T17:45:08+00:00
<li><b>votes</b>: 1
<li><b>watchers</b>: 3
<li><b>imported</b>: 2025-12-06
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>I think the idea of a declarative (and not imperative) pipeline is a lot easier on the writers end (and I assume the implementers end as well), but I think the implementation of several serial stages each fanning out to a number of parallel stages is limiting to the overall description of the dependencies between different steps/nodes. In my case (simplifying, please excuse pseudo-code), I'd like to do something like this:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">stage(<span class="code-quote">"checkout"</span>) { node () { git XXX } }
stage(<span class="code-quote">"build"</span>) {
    parallel (<span class="code-quote">"ubuntu"</span>) { node(<span class="code-quote">"ubuntu"</span>) { make XXX }},
                 (<span class="code-quote">"centos"</span>) { node(<span class="code-quote">"centos"</span>) { make YYY }},
                 <span class="code-comment">// tons more distributions here
</span>}
stage (<span class="code-quote">"test"</span>) {
    node (<span class="code-quote">"ubuntu"</span>) { test-ubuntu ZZZ }
}

stage (<span class="code-quote">"publish"</span>) { node() { publish.sh }
</pre>
</div></div>

<p>Since I know the "test" stage only depends on the "ubuntu" part of the build stage, I'd like the test stage to fire without waiting on the whole "build" stage completes. From what I see, the parallel block waits for everything inside to complete, but I know the dependencies and can speed it up (in my case, there are a lot more nodes who match the testing than who don't). If I were to write it in bash, I would do something like:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">git checkout XXX
build ubuntu &amp;
TEST_PID=$!
build centos &amp;
wait $TEST_PID
test-ubuntu
wait # wait <span class="code-keyword">for</span> all testing and bulding to complete
publish
</pre>
</div></div>

<p>This is a rough way to explain the dependency graph to bash by saying "wait on 'build ubuntu' before starting the testing, but centos can take as long as it wants" and "before going to the publish step, make sure everything else is done". In my case, all the build steps for different distributions take forever because there's a finite number of VMs that can be spun up, but there's an abundance of "ubuntu" nodes that can be used for testing and compiling. It slows things down a ton to have to wait for the entire parallel node to finish before moving on, when there are free executors in latter steps that can start running.</p></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> abayer:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/39932?focusedId=278272&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-278272">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-39932?focusedId=278272&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-278272">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>fwiw, I have plans for a stage dependency graph execution model as part of Declarative post-1.0, where you'd specify your stages and which other stages they depend on, and the execution would figure out the appropriate ordering based on that, including smart parallelization etc.</p>
</pre>
</details>

</ul>
</details>

<p>fwiw, I have plans for a stage dependency graph execution model as part of Declarative post-1.0, where you'd specify your stages and which other stages they depend on, and the execution would figure out the appropriate ordering based on that, including smart parallelization etc.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> abayer:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/39932?focusedId=323110&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-323110">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-39932?focusedId=323110&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-323110">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>We're not going to actually be doing dependency ordering, but in the upcoming 1.3, you'll be able to do sequential lists of stages in parallel with each other (see <a href="https://issue-redirect.jenkins.io/issue/46809" title="Allow sequential stages inside parallel in Declarative syntax" class="issue-link" data-issue-key="JENKINS-46809"><del>JENKINS-46809</del></a>), which I think will do the trick for you.</p>
</pre>
</details>

</ul>
</details>

<p>We're not going to actually be doing dependency ordering, but in the upcoming 1.3, you'll be able to do sequential lists of stages in parallel with each other (see <a href="https://issue-redirect.jenkins.io/issue/46809" title="Allow sequential stages inside parallel in Declarative syntax" class="issue-link" data-issue-key="JENKINS-46809"><del>JENKINS-46809</del></a>), which I think will do the trick for you.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/bitwiseman.png" title="bitwiseman's avatar" /> bitwiseman:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/39932?focusedId=378287&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-378287">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-39932?focusedId=378287&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-378287">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Bulk closing resolved issues. </p>
</pre>
</details>

</ul>
</details>

<p>Bulk closing resolved issues. </p>
